### ISSUE TITLE
[ASTextNode] Center-aligned text not rendering correctly 

### ISSUE THREAD

**Issue by GitGadd** [facebook/AsyncDisplayKit#2677]
After updating from 2.0-beta.1 to 2.0-beta.2, center-aligned text in our custom alerts is rendering incorrectly. Note that using 'left' text alignment and taking the same screenshot looks OK.

Update: Manually setting the style.width of each ASTextNode to the expected width seems to fix it as a work-around, but would obviously break down if the constrained width was a range.

Screenshot of issue:

![screen shot 2016-11-29 at 08 41 08](https://cloud.githubusercontent.com/assets/3338062/20700785/a0160550-b60f-11e6-98cc-2048a8ed013c.png)

Attributed text creation:
```
titleAttributes[NSParagraphStyleAttributeName] = {
    let para = NSMutableParagraphStyle()
    para.alignment = .center
    return para
}()
titleNode.attributedText = NSAttributedString(string: title, attributes: titleAttributes)
```

Layout:
```
fileprivate override func layoutSpecThatFits(_ constrainedSize: ASSizeRange) -> ASLayoutSpec {
		var children = [titleNode]
		if let messageNode=messageNode {
			children.append(messageNode)
		}
		return ASBackgroundLayoutSpec(
			child: ASInsetLayoutSpec(
				insets: UIEdgeInsets(horizontal: AXTheme.defaultMargin*2, vertical: AXTheme.defaultMargin*3),
				child: ASStackLayoutSpec(direction: .vertical,
					spacing: AXTheme.defaultMargin,
					justifyContent: .start,
					alignItems: .center,
					children: children)),
			background: backgroundImageNode)
}
```

**Comment by hannahmbanana** [comment]
I tried reproducing this in the `LayoutSpecExamples-Swift` project, but was unable to reproduce the bug. Here's the diff:  [GaddLayoutBug.txt](https://github.com/facebook/AsyncDisplayKit/files/620100/GaddLayoutBug.txt)

@GitGadd - could you print out the ascii art string as well as the full hierarchy (using recursiveDescription)?

**Comment by GitGadd** [comment]
Only the message part (title+text):

```
"---------------ASBackgroundLayoutSpec---------------
|  --------ASInsetLayoutSpec--------               |
|  |  -----ASStackLayoutSpec-----  |               |
|  |  |        ASTextNode       |  |               |
|  |  |        ASTextNode       |  |  ASImageNode  |
|  |  ---------------------------  |               |
|  ---------------------------------               |
----------------------------------------------------"
```

The entire alert:

```
"--------------------------------------------ASStackLayoutSpec--------------------------------------------
|  -----------------------------------------ASStackLayoutSpec-----------------------------------------  |
|  |  _TtC4AuxyP33_8E4AE96FB573DD15CA6885CFCEBC828646AXActionPanelControllerTextLabelWithBackground  |  |
|  |                                           ASImageNode                                           |  |
|  |                     -------------------ASStackLayoutSpec-------------------                     |  |
|  |                     |  Auxy.AXButtonNode  ASImageNode  Auxy.AXButtonNode  |                     |  |
|  |                     -------------------------------------------------------                     |  |
|  |                                     --ASAbsoluteLayoutSpec--                                    |  |
|  |                                     |     ASLayoutSpec     |                                    |  |
|  |                                     ------------------------                                    |  |
|  ---------------------------------------------------------------------------------------------------  |
---------------------------------------------------------------------------------------------------------"
```

Recursive description of alert:

```
"<<_TtCC4Auxy23AXActionPanelControllerP33_8E4AE96FB573DD15CA6885CFCEBC82864Node: 0x7fcea1e31a20> alpha:1.00 isLayerBacked:0 frame:{{0, 0}, {375, 667}} (null)>
 | <<_TtC4AuxyP33_8E4AE96FB573DD15CA6885CFCEBC828646AXActionPanelControllerTextLabelWithBackground: 0x7fcea1d35c30> alpha:1.00 isLayerBacked:0 frame:{{0, 0}, {0, 0}} (null)>
 |  | <<ASImageNode: 0x7fcea28a7a00> alpha:1.00 isLayerBacked:0 frame:{{0, 0}, {0, 0}} (null)>
 |  | <<ASTextNode: 0x7fcea296bc00; text = \"Are you sure?\"> alpha:1.00 isLayerBacked:0 frame:{{0, 0}, {0, 0}} (null)>
 |  | <<ASTextNode: 0x7fcea2878c00; text = \"Deleting an instrument can\'t be undone.\"> alpha:1.00 isLayerBacked:0 frame:{{0, 0}, {0, 0}} (null)>
 | <<Auxy.AXButtonNode: 0x7fcea2879800> alpha:1.00 isLayerBacked:0 frame:{{0, 0}, {0, 0}} (null)>
 | <<Auxy.AXButtonNode: 0x7fcea2870000> alpha:1.00 isLayerBacked:0 frame:{{0, 0}, {0, 0}} (null)>
 | <<ASImageNode: 0x7fcea2955000> alpha:1.00 isLayerBacked:0 frame:{{0, 0}, {0, 0}} (null)>
 | <<ASImageNode: 0x7fcea2956600> alpha:1.00 isLayerBacked:0 frame:{{0, 0}, {0, 0}} (null)>
"
```

The node is the root node of an ASViewController being presented through the regular -[UIViewController presentViewController:animated:completion:] api

**Comment by GitGadd** [comment]
Attaching a project that reproduces the issue. It seems to happen when presented through UIKit in a contentSize that is less than the hosting view's size, causing the node to resize multiple times during the various parts in UIKits layout process.

There was also an offending flexBasis being set in our code that I'm not even sure why it was there in the first place, which seems to have no effect except causing this issue. That line has been added (and commented) to this sample project to recreate the issue as close to the real case as possible.

[ASTextNodeAlignmentIssue.zip](https://github.com/facebook/AsyncDisplayKit/files/620245/ASTextNodeAlignmentIssue.zip)

**Comment by hannahmbanana** [comment]
Punting this to 2.0.1 unless @maicki believes it is more urgent.

**Comment by maicki** [comment]
@GitGadd Hey, I looked a bit into it and I would be great if you could elaborate a bit more about the issue ... as setting `flexB@GitGadd Hey, I looked a bit into it and I would be great if you could elaborate a bit more about the issue ... as setting `flexBasis` to 300 pt means the stack will have a minimum width of 300 so based on this value it will layout all of the subnodes based. Furthermore if the hosting view is smaller than the 300 it will still use the 300 to layout all of the children.

I think there are multiple ways to move change the layout description in case you would like it to have like this:


Change `layoutSpecThatFits:` in `ViewController` to this:

```Swift
override func layoutSpecThatFits(_ constrainedSize: ASSizeRange) -> ASLayoutSpec {
	
	let stack = ASStackLayoutSpec(direction: .vertical, spacing: 0, justifyContent: .start, alignItems: .stretch, children:[messageNode])
	
         // The max width you would like to have. It should get passed down based on your
         // container view
        stack.style.width = ASDimensionMake(constrainedSize.max.width)
	
	
	let spec = ASStackLayoutSpec(direction: .horizontal, spacing: 0, justifyContent: .center, alignItems: .center, children: [stack])
	return spec
}
```

Or if you don't want to set the `width` static to the stack you could solve it by changing `layoutSpecThatFits` in `ViewController` to:

```Swift
override func layoutSpecThatFits(_ constrainedSize: ASSizeRange) -> ASLayoutSpec {
	
	let stack = ASStackLayoutSpec(direction: .vertical, spacing: 0, justifyContent: .start, alignItems: .stretch, children:[messageNode])
    stack.style.flexShrink = 1.0	
	
	let spec = ASStackLayoutSpec(direction: .horizontal, spacing: 0, justifyContent: .center, alignItems: .center, children: [stack])
	return spec
}
```

and `layoutSpecThatFits:` in `AXActionPanelControllerTextLabelWithBackground`:

```Swift
fileprivate override func layoutSpecThatFits(_ constrainedSize: ASSizeRange) -> ASLayoutSpec {
	var children = [titleNode]
	if let messageNode = messageNode {
        
                messageNode.style.flexShrink = 1.0
		
		children.append(messageNode)
	}
	let horizontalMargin:CGFloat = 20
	
	let spec = ASBackgroundLayoutSpec(
		child: ASInsetLayoutSpec(
			insets: UIEdgeInsetsMake(30, horizontalMargin, 30, horizontalMargin),
			child: ASStackLayoutSpec(direction: .vertical,
			                         spacing: 10,
			                         justifyContent: .start,
			                         alignItems: .center,
			                         children: children)),
		background: backgroundImageNode)
	return spec
}
```

Hope that helps ...asis` to 300 pt means the stack will have a minimum width of 300 so based on this value it will layout all of the subnodes based. Furthermore if the hosting view is smaller than the 300 it will still use the 300 to layout all of the children.

**Comment by GitGadd** [comment]
Thanks Maicki!

I really appreciate you digging into this, but the reason I posted the project to reproduce the issue was in case somebody wanted to dig into the underlying issue. The issues in our app are already resolved (by removing the flexBasis line). 

I can't really elaborate on the flexBasis as I don't know why it was there in the first place, but the weird thing is that the flexBasis was being set on the vertical stack, not the horizontal one.

**Comment by maicki** [comment]
The question is what is the underlying issue you are referring? If you set a `flexBasis` to 300 it's expected that the stack will have a min width of 300 that will be used to layout it's children...

**Comment by GitGadd** [comment]
Yes, but programmig errors aside, the text node is drawing it's contents way outside it's bounds. Look at the original screenshot in the first post; the bounds are outlined by Xcode and are the expected size and position, while the contents continue outside.

**Comment by rewandy** [comment]
I am encountering the same issue with center-aligned `ASTextNode`s that have been sized with `flexBasis`. The final frame of my text nodes is correct, but the text is not being centered within them.

#### Before

<img width="371" alt="screen shot 2016-12-04 at 4 58 35 pm" src="https://cloud.githubusercontent.com/assets/4761768/20870800/02f8b302-ba43-11e6-851d-4f2048b10a57.png">

#### After

<img width="372" alt="screen shot 2016-12-04 at 4 51 56 pm" src="https://cloud.githubusercontent.com/assets/4761768/20870804/0c6339c6-ba43-11e6-88b3-22751e21250d.png">

<hr>

I need to use `flexBasis` so I can tell each "box" to take up `25%` of the available space horizontally. Like I said, the boxes **are** sized correctly, it's just that the `ASTextNode`s are not rendering their text as expected.

This appears to have regressed sometime between 4d16d5840735cf0a06a258bcb4ce4f0c714e593f and 2.0 rc1.

**Comment by GitGadd** [comment]
Saw this commit earlier when reading through changes; Perhaps the new implementation is caching a size from one of the previous layout passes, and then not recomputing when a later pass invalidates it?

In my case: The alert is first being laid out at full screen width, then later as the smaller alert size. The text is then drawn at what seems to be full width anyway.

In @rewandy 's case: The nodes are first being laid out at their intrinsic size (i.e. rendering centered doesn't add padding), then later laid out at 25% of the stack width.

**Comment by rewandy** [comment]
[This](https://github.com/facebook/AsyncDisplayKit/pull/2643) PR fixes my particular issue (thanks @maicki!), as mentioned in the Slack channel. But it sounds like the team needs to do further testing before they can land it. 

@GitGadd maybe you can give that a shot as well? Though it sounds like you may have already worked around the issue.

**Comment by GitGadd** [comment]
@maicki The mentioned PR does resolve the issue in my sample project as well!

**Comment by Adlai-Holler** [comment]
Resolved in #2643.
